### PR TITLE
Get products endpoint

### LINKED ISSUE #2: Get products endpoint
Develop the get products endpoint as described here

https://thepos.stoplight.io/docs/thepos/b3A6MjMwNDAyNzU-get-products

### PR DESCRIPTION
**Pull request by devmsh**
Developing the first endpoint of the POS system to retrieve the basic products list. It creates the `products` table with the required columns as described in the [API spec](https://thepos.stoplight.io/docs/thepos/b3A6MjMwNDAyNzU-get-products)

I decided to use `in-memory SQlite` database for the testing environment and tried as possible to use the Laravel standards when applicable.

I changed the default `RouteServiceProvider` to make the API versioning support easier in the future, where v1 mapped to the default `api.php` file

Some API spec was changed to comply with the Laravel standard and it will be reflected to the SPEC once the PR accepted

Closes #2

### CODE REVIEW

File: database/migrations/2021_10_10_131533_create_products_table.php [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+class CreateProductsTable extends Migration
+{
+    public function up()
+    {
+        Schema::create('products', function (Blueprint $table) {
+            $table->id();
+            $table->string('sku');
+            $table->string('name');
+            $table->integer('price');
+            $table->float('tax_rate');
```
Comment (fahjan) [inline database/migrations/2021_10_10_131533_create_products_table.php]: i prefer to use decimal data type. 
as i worked before, decimal type solved many issues for financials programs as a "round".
may be use:
$table->decimal('price', 10, 2);

Comment (fahjan) [reply database/migrations/2021_10_10_131533_create_products_table.php]: also 
price, tax_rate and taxed_price

File: app/Http/Resources/ProductResource.php [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+namespace App\Http\Resources;
+
+use Illuminate\Http\Resources\Json\JsonResource;
+
+class ProductResource extends JsonResource
+{
+    public function toArray($request)
+    {
+        return parent::toArray($request);
```
Comment (fahjan) [inline app/Http/Resources/ProductResource.php]: i think that you should return only needed fields, not all fields from query.
and if you need created_at, updated_at, you should return a human readable dateTime.

Comment (MostafaBalata) [reply app/Http/Resources/ProductResource.php]: I completely agree with @fahjan otherwise what is the benefit from it?

**Review by MostafaBalata — COMMENTED** [review]
- Write PR description.
- If the PR is not finished yet prefix the title with [WIP]

File: routes/api.php [hunk 3]
```diff
@@ -17,3 +17,5 @@
 Route::middleware('auth:sanctum')->get('/user', function (Request $request) {
     return $request->user();
 });
+
+Route::apiResource('v1/products', \App\Http\Controllers\ProductController::class);
```
Comment (ahmdqader) [inline routes/api.php]: Put v1 in route's group, so you don't need to repeat it in every route, make V1 Subdirectory inside controllers to separate every version controllers.

File: app/Http/Resources/ProductResource.php (hunk 2 above)
Comment (devmsh) [reply app/Http/Resources/ProductResource.php]: Totally agreed about returning the required data manually, yet it's the front-end task to format the DateTime the way they want

File: database/migrations/2021_10_10_131533_create_products_table.php (hunk 1 above)
Comment (devmsh) [reply database/migrations/2021_10_10_131533_create_products_table.php]: 👍

File: routes/api.php (hunk 3 above)
Comment (devmsh) [reply routes/api.php]: I will move it entirely to the `RouteServiceProvider`, so we can use a different files for each version (e.g. api_v2.php) in the future.

**Comment by 123mousab** [comment]
Create standard class for return response data for all classes.

**Comment by devmsh** [comment]
@123mousab while we use API Resource, I think no need to do so.

Can you please give me more details if you think we still need that?

**Comment by nidalb** [comment]
use $scale = 2 to decimal digits

**Comment by ahmdqader** [comment]
> @123mousab while we use API Resource, I think no need to do so.
> 
> Can you please give me more details if you think we still need that?

Creating generalized response builder helps to manage complex api structure in a predictable way for the mobile developers, this will be necessary in case you used paginations, status messages, and returning multiple types of data in the same request.

File: app/Http/Resources/ProductResource.php (hunk 2 above)
Comment (fahjan) [reply app/Http/Resources/ProductResource.php]: i think its a backend task, imagine if you have a 3 separated systems, for android, ios and desktop. three systems will handle date formate ?
another case, if you need to display date as a user country format D-M-Y or M-D-Y, or Hijri date.

**Comment by devmsh** [comment]
@ahmdqader can you please give me an example? Api Resource class already do that, why I need extra class or code?

File: database/migrations/2021_10_10_131533_create_products_table.php (hunk 1 above)
Comment (AboSaleem) [reply database/migrations/2021_10_10_131533_create_products_table.php]: Why we do not have a separate table for taxes?
if the tax changed we need to update all products!

**Comment by devmsh** [comment]
> use $scale = 2 to decimal digits

It's already the default Laravel decimal digitis

```php
unsignedDecimal($column, $total = 8, $places = 2)
```

**Comment by 123mousab** [comment]
> @ahmdqader can you please give me an example? Api Resource class already do that, why I need extra class or code?



> > @123mousab while we use API Resource, I think no need to do so.
> > Can you please give me more details if you think we still need that?
> 
> Creating generalized response builder helps to manage complex api structure in a predictable way for the mobile developers, this will be necessary in case you used paginations, status messages, and returning multiple types of data in the same request.

not problem this only to manage structure data and handle messages in statuses true and failed with return errors in the errors place in the cases of error.

**Pull request merged**